CI: BalanceConfigurationTests.testReplicaBalance fails in a reproducible way

Going to mute the test.

## Reproduction line 
**does** reproduce
```
./gradlew :server:test \
  -Dtests.seed=DDBE7ADDD8F01971 \
  -Dtests.class=org.elasticsearch.cluster.routing.allocation.BalanceConfigurationTests \
  -Dtests.method="testReplicaBalance" \
  -Dtests.security.manager=true \
  -Dtests.locale=cs \
  -Dtests.timezone=Australia/Brisbane \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

## Example relevant log:
```
23:35:47   2> REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=DDBE7ADDD8F01971 -Dtests.class=org.elasticsearch.cluster.routing.allocation.BalanceConfigurationTests -Dtests.method="testReplicaBalance" -Dtests.security.manager=true -Dtests.locale=cs -Dtests.timezone=Australia/Brisbane -Dcompiler.java=11 -Druntime.java=8
23:35:47   1> [2018-11-08T07:35:47,314][INFO ][o.e.c.r.a.BalanceConfigurationTests] [testReplicaBalance] complete rebalancing
23:35:47   2> NOTE: leaving temporary files on disk at: /var/lib/jenkins/workspace/elastic+elasticsearch+master+intake/server/build/testrun/test/J0/temp/org.elasticsearch.cluster.routing.allocation.BalanceConfigurationTests_DDBE7ADDD8F01971-001
23:35:47   1> [2018-11-08T07:35:47,315][INFO ][o.e.c.r.a.BalanceConfigurationTests] [testReplicaBalance] after test
23:35:47   2> NOTE: test params are: codec=Asserting(Lucene80): {}, docValues:{}, maxPointsInLeafNode=1291, maxMBSortInHeap=7.145509876487136, sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@74dca44f), locale=cs, timezone=Australia/Brisbane
23:35:47 FAILURE 0.21s J0 | BalanceConfigurationTests.testReplicaBalance <<< FAILURES!
23:35:47   2> NOTE: Linux 4.4.156-94.64-default amd64/Oracle Corporation 1.8.0_192 (64-bit)/cpus=16,threads=1,free=362555832,total=519569408
23:35:47    > Throwable #1: java.lang.AssertionError: 
23:35:47   2> NOTE: All tests run in this JVM: [NodeInfoStreamingTests, CountDownTests, FileInfoTests, QueryRescoreModeTests, KeyStoreWrapperTests, RetryTests, ClusterApplierServiceTests, FieldMemoryStatsTests, DoubleTermsTests, LuceneChangesSnapshotTests, SignificantLongTermsTests, RootObjectMapperTests, RestIndicesActionTests, BulkByScrollTaskStatusOrExceptionTests, BalanceConfigurationTests]
23:35:47    > Expected: a value equal to or greater than <4>
23:35:47    >      but: <3> was less than <4>
23:35:47    > 	at __randomizedtesting.SeedInfo.seed([DDBE7ADDD8F01971:B47081853A0CDBC0]:0)
23:35:47    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
23:35:47    > 	at org.elasticsearch.cluster.routing.allocation.BalanceConfigurationTests.assertReplicaBalance(BalanceConfigurationTests.java:209)
23:35:47    > 	at org.elasticsearch.cluster.routing.allocation.BalanceConfigurationTests.testReplicaBalance(BalanceConfigurationTests.java:112)
23:35:47    > 	at java.lang.Thread.run(Thread.java:748)
```

## Frequency

6 times this week on an upwards trend.

--- a/server/src/test/java/org/elasticsearch/cluster/routing/allocation/BalanceConfigurationTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/allocation/BalanceConfigurationTests.java
@@ -45,8 +45,11 @@ import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.test.gateway.TestGatewayAllocator;
 import org.hamcrest.Matchers;
 
+import java.util.stream.Collectors;
+
 import static org.elasticsearch.cluster.routing.ShardRoutingState.INITIALIZING;
 import static org.elasticsearch.cluster.routing.ShardRoutingState.STARTED;
+import static org.elasticsearch.cluster.routing.ShardRoutingState.UNASSIGNED;
 
 public class BalanceConfigurationTests extends ESAllocationTestCase {
 
@@ -85,34 +88,32 @@ public class BalanceConfigurationTests extends ESAllocationTestCase {
             (numberOfNodes + 1) - (numberOfNodes + 1) / 2, numberOfIndices, numberOfReplicas, numberOfShards, balanceTreshold);
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/35365")
     public void testReplicaBalance() {
         /* Tests balance over replicas only */
         final float indexBalance = 0.0f;
         final float replicaBalance = 1.0f;
-        final float balanceTreshold = 1.0f;
+        final float balanceThreshold = 1.0f;
 
         Settings.Builder settings = Settings.builder();
         settings.put(ClusterRebalanceAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ALLOW_REBALANCE_SETTING.getKey(),
             ClusterRebalanceAllocationDecider.ClusterRebalanceType.ALWAYS.toString());
         settings.put(BalancedShardsAllocator.INDEX_BALANCE_FACTOR_SETTING.getKey(), indexBalance);
         settings.put(BalancedShardsAllocator.SHARD_BALANCE_FACTOR_SETTING.getKey(), replicaBalance);
-        settings.put(BalancedShardsAllocator.THRESHOLD_SETTING.getKey(), balanceTreshold);
+        settings.put(BalancedShardsAllocator.THRESHOLD_SETTING.getKey(), balanceThreshold);
 
         AllocationService strategy = createAllocationService(settings.build(), new TestGatewayAllocator());
 
         ClusterState clusterState = initCluster(strategy);
-        assertReplicaBalance(logger, clusterState.getRoutingNodes(), numberOfNodes, numberOfIndices,
-            numberOfReplicas, numberOfShards, balanceTreshold);
+        assertReplicaBalance(clusterState.getRoutingNodes(), numberOfNodes, numberOfIndices,
+            numberOfReplicas, numberOfShards, balanceThreshold);
 
         clusterState = addNode(clusterState, strategy);
-        assertReplicaBalance(logger, clusterState.getRoutingNodes(), numberOfNodes + 1,
-            numberOfIndices, numberOfReplicas, numberOfShards, balanceTreshold);
+        assertReplicaBalance(clusterState.getRoutingNodes(), numberOfNodes + 1,
+            numberOfIndices, numberOfReplicas, numberOfShards, balanceThreshold);
 
         clusterState = removeNodes(clusterState, strategy);
-        assertReplicaBalance(logger, clusterState.getRoutingNodes(),
-            (numberOfNodes + 1) - (numberOfNodes + 1) / 2, numberOfIndices, numberOfReplicas, numberOfShards, balanceTreshold);
-
+        assertReplicaBalance(clusterState.getRoutingNodes(),
+            numberOfNodes + 1 - (numberOfNodes + 1) / 2, numberOfIndices, numberOfReplicas, numberOfShards, balanceThreshold);
     }
 
     private ClusterState initCluster(AllocationService strategy) {
@@ -199,9 +200,25 @@ public class BalanceConfigurationTests extends ESAllocationTestCase {
     }
 
 
-    private void assertReplicaBalance(Logger logger, RoutingNodes nodes, int numberOfNodes, int numberOfIndices, int numberOfReplicas,
-                                      int numberOfShards, float treshold) {
-        final int numShards = numberOfIndices * numberOfShards * (numberOfReplicas + 1);
+    private void assertReplicaBalance(RoutingNodes nodes, int numberOfNodes, int numberOfIndices, int numberOfReplicas,
+        int numberOfShards, float treshold) {
+        final int unassigned = nodes.unassigned().size();
+
+        if (unassigned > 0) {
+            // Ensure that if there any unassigned shards, all of their replicas are unassigned as well
+            // (i.e. unassigned count is always [replicas] + 1 for each shard unassigned shardId)
+            nodes.shardsWithState(UNASSIGNED).stream().collect(
+                Collectors.toMap(
+                    ShardRouting::shardId,
+                    s -> 1,
+                    (a, b) -> a + b
+            )).values().forEach(
+                count -> assertEquals(numberOfReplicas + 1, count.longValue())
+            );
+        }
+        assertEquals(numberOfNodes, nodes.size());
+
+        final int numShards = numberOfIndices * numberOfShards * (numberOfReplicas + 1) - unassigned;
         final float avgNumShards = (float) (numShards) / (float) (numberOfNodes);
         final int minAvgNumberOfShards = Math.round(Math.round(Math.floor(avgNumShards - treshold)));
         final int maxAvgNumberOfShards = Math.round(Math.round(Math.ceil(avgNumShards + treshold)));

